libraries/from-bodyxml/index.js

```diff
@@ -40,371 +51,385 @@ function toValidLayoutWidth(layoutWidth) {
+
+      let transformer =
+        xmlnode.name == "content" || xmlnode.name == "ft-content"
+          ? String(xmlnode.attributes.type)
+          : xmlnode.name;
 
-			let transformer =
-				(xmlnode.name == "content" || xmlnode.name == "ft-content")
-					? String(xmlnode.attributes.type)
-					: xmlnode.name
+      if (transformer in transformers) {
+        let ctnode = transformers[transformer](xmlnode);
+        if (ctnode.type === "__LIFT_CHILDREN__") {
+          // we don't want this node to stick around, but we want to keep its' children
+          return xmlnode.children.flatMap(walk);
+        } else if ("children" in ctnode && ctnode.children === null) {
+          // this is how we indicate we shouldn't iterate, but this thing
+          // shouldn't have any children
+          delete ctnode.children;
+          return ctnode;
+        } else if ("children" in ctnode && Array.isArray(ctnode.children)) {
+          return ctnode;
+        } else if ("children" in xmlnode) {
+          return {
+            ...ctnode,
+            // this is a flatmap because of <experimental/>
+            children: xmlnode.children.flatMap(walk),
+          };
+        }
+        return ctnode;
+      } else {
+        return { type: "__UNKNOWN__" };
+      }
+    } else if (isXText(xmlnode)) {
+      return {
+        type: "text",
+        value: xmlnode.value,
+      };
+    } else {
+      return { type: "__UNKNOWN__" };
+    }
+  })(bodyxast);
+}
 
-			if (transformer in transformers) {
-				let ctnode = transformers[transformer](xmlnode)
-				if(ctnode.type === "__LIFT_CHILDREN__") {
-					// we don't want this node to stick around, but we want to keep its' children
-					return xmlnode.children.flatMap(walk);
-				} else if ("children" in ctnode && ctnode.children === null) {
-					// this is how we indicate we shouldn't iterate, but this thing
-					// shouldn't have any children
-					delete ctnode.children
-					return ctnode
-				} else if ("children" in ctnode && Array.isArray(ctnode.children)) {
-					return ctnode
-				} else if ("children" in xmlnode) {
-					return {
-						...ctnode,
-						// this is a flatmap because of <experimental/>
-						children: xmlnode.children.flatMap(walk),
-					}
-				}
-				return ctnode
-			} else {
-				return {type: "__UNKNOWN__"}
-			}
-		} else if (isXText(xmlnode)) {
-			return {
-				type: "text",
-				value: xmlnode.value,
-			}
-		} else {
-			return {type: "__UNKNOWN__"}
-		}
-	})(bodyxast)
+/**
+ * Turns e.g. </p> </body> into </p></body>, which a lot of our content has
+ * This prevents random strat text nodes being created in bodyTree
+ * @param {string} xml
+ * @returns {string}
+ */
+function removeWhitespaceBeforeBodyTag(xml) {
+  return xml.replace("</p> </body>", "</p></body>");
 }
 
 /** @param {string} bodyxml */
 export function fromXML(bodyxml) {
-	return fromXast(xastFromXml(bodyxml))
+  return fromXast(xastFromXml(removeWhitespaceBeforeBodyTag(bodyxml)));
 }
```

Comment:
This is the bit that has actually changed (sorry my editor formatted everything and I CBA to split it out 🙈 )